src/constructor_parser.rs

```diff
@@ -0,0 +1,380 @@
+// Copyright 2018-2021 the Deno authors. All rights reserved. MIT license.
+
+use crate::tokenizer::Token;
+use crate::tokenizer::TokenType;
+use crate::ParseError;
+use crate::UrlPatternInit;
+
+// Ref: https://wicg.github.io/urlpattern/#constructor-string-parser-state
+#[derive(Eq, PartialEq)]
+enum ConstructorStringParserState {
+  Init,
+  Protocol,
+  Authority,
+  Username,
+  Password,
+  Hostname,
+  Port,
+  Pathname,
+  Search,
+  Hash,
+  Done,
+}
+
+// Ref: https://wicg.github.io/urlpattern/#constructor-string-parser
+struct ConstructorStringParser<'a> {
+  input: &'a str,
+  token_list: Vec<Token>,
+  result: UrlPatternInit,
+  component_start: usize,
+  token_index: usize,
+  token_increment: usize,
+  group_depth: usize,
+  should_treat_as_standard_url: bool,
+  state: ConstructorStringParserState,
+}
+
+impl<'a> ConstructorStringParser<'a> {
+  // Ref: https://wicg.github.io/urlpattern/#rewind
+  #[inline]
+  fn rewind(&mut self) {
+    self.token_index = self.component_start;
+    self.token_increment = 0;
+  }
+
+  // Ref: https://wicg.github.io/urlpattern/#is-a-hash-prefix
+  #[inline]
+  fn is_hash_prefix(&self) -> bool {
+    self.is_non_special_pattern_char(self.token_index, "#")
+  }
+
+  // Ref: https://wicg.github.io/urlpattern/#is-a-protocol-suffix
+  #[inline]
+  fn is_protocol_suffix(&self) -> bool {
+    self.is_non_special_pattern_char(self.token_index, ":")
+  }
+
+  // Ref: https://wicg.github.io/urlpattern/#is-a-search-prefix
+  fn is_search_prefix(&self) -> bool {
+    if self.is_non_special_pattern_char(self.token_index, "?") {
+      return true;
+    }
+    if self.token_list[self.token_index].value != "?" {
+      return false;
+    }
+    let previous_index = self.token_index - 1;
+    #[allow(clippy::absurd_extreme_comparisons)]
+    if previous_index < 0 {
+      // TODO: can self.token_index ever be negative?
```

Comment:
not if you are using a `usize` (unsigned)

Reply:
I am aware, just thinking in terms of spec. frm what i can see from the spec though, i dont see that being the case, but i might have missed something